### DISCUSSION TITLE
OpenShift: 1337 is not an allowed group

### DISCUSSION THREAD

**Discussion by emilbillberg** [fluxcd/flux2#1051] [Q&A] [answered]
Hi! 

I am running Openshift version `4.5.0-0.okd-2020-10-15-235428` which is running kubernetes version `v1.18.3` behind the hood. 

I ran this bootstrap command:
```sh
flux bootstrap github \
  --owner=<org> \
  --repository=<repo> \
  --branch=<branch> \
  --path=./clusters/<cluster-name> \
  --components-extra=image-reflector-controller,image-automation-controller \
  --token-auth \
  --toleration-keys <keys> \
  --version v0.9.0 \
  --private
```

However, the bootstrap fails to create pods for components `image-automation-controller`, `image-reflector-controller`, and `source-controller`. The message is:
```
pods <component-name> is forbidden: unable to validate against any security context constraint:[fsGroup: Invalid value: []int64{1337}: 1337 is not an allowed group]
```

What is the recommended approach to circumvent this?

**Comment by hiddeco** [comment]
https://github.com/weaveworks/flux2-openshift, available at https://operatorhub.io/operator/flux should contain the required patches to make this work properly for OpenShift.

**Reply by emilbillberg** [reply]
thank you! I will take a look 👍

**Comment by saada** [comment, marked as answer]
Flux recently added a feature that relies on the securityContext.fsGroup pod spec. In order for that to work on Openshift, we need to add docs for a pre-requisites step to create an Openshift SecurityContextConstraint to give permission to Flux to use this fsGroup. We are working on adding documentation with these steps. We'll report back to this discussion when it's available.

Thanks for reporting

**Reply by emilbillberg** [reply]
Oh great! Very much appreciated!

Thanks for great work and a great project!

**Reply by saada** [reply]
@emilbillberg I have the draft for the doc addition below. Can you try the below steps to see if it works?

## Pre-Requisites

Before installing Flux, perform the following steps to bind Flux to the `privileged` SecurityContextConstraint (SCC).

```sh
# replace default "flux-system" with the project/namespace name for your flux installation
namespace="flux-system"
oc adm policy add-scc-to-user privileged system:serviceaccount:$namespace:source-controller
oc adm policy add-scc-to-user privileged system:serviceaccount:$namespace:kustomize-controller
oc adm policy add-scc-to-user privileged system:serviceaccount:$namespace:image-automation-controller
oc adm policy add-scc-to-user privileged system:serviceaccount:$namespace:image-reflector-controller
```

If you want to narrow down permission scope even further than `privileged` (SCC), you need to create an SCC with only fsGroup permissions as shown below and bind it to the same service accounts above.

```yaml
kind: SecurityContextConstraints
apiVersion: v1
metadata:
  name: scc-flux
fsGroup:
  type: RunAsAny
```

**Reply by emilbillberg** [reply]
Thank you! I haven't had the chance to try yet but will let you know as soon as I have

**Reply by pestebogdan** [reply]
I can confirm this works on OCP 4.6.17 . Created the "flux-system" namespace first, then the 4 add-scc-to-user and finally the flux bootstrap command. NOTE: You may have to run the `flux uninstall` first if you tried before doing the NS creation and SCC rules and start fresh. Thanks a lot @saada  .

**Reply by saada** [reply]
Awesome. PR for updated docs can be found here: https://github.com/weaveworks/flux2-openshift/pull/3. Let's mark this question as "answered" cc: @hiddeco .

The docs should show up in https://operatorhub.io/operator/flux soon

**Reply by emilbillberg** [reply]
I can verify that it works too 👍

**Reply by dontlaugh** [reply]
The docs still could use some love. https://fluxcd.io/flux/use-cases/openshift/#flux-installation-with-cli

They still mention the old Kustomization spec, iiuc.

Anyway, I needed to do this: create the `flux-system` namespace _first_, then apply the "privileged" SCCs mentioned above, and then install flux components. Essentially @pestebogdan 's answer.

I'm going to use "privileged" to get my deployment unstuck, and I'll try narrowing the permissions later.

**Reply by kingdonb** [reply]
@dontlaugh Please take a look at https://operatorhub.io/operator/flux

There is where I believe the detail you were looking for is documented, but I'm frankly not sure how many people that are installing Flux on OpenShift will look there if they've already decided to use "Generic" (non-OLM-driven) flux. Can you say where you would have looked to find this information, if you didn't find it there?

Please open an issue or PR with some suggestions if you can find what you'd like to see changed, or ping me on CNCF Slack with a DM or thread for your suggestions and I will follow them up after KubeCon 👍
